File: demo/Fortran/2D/Ellipse/ellipse.f90
!* -------------------------------------------------------------------------- *
!* DESCRIPTION (reference phase where f(x,y) < 0):                            *
!* ellipse inside the square [0,1]x[0,1]                                      *
!* f(x,y) = c1*x^2 + c2*y^2 + c3*x*y + c4*x + c5*y - c6                       *
!* PARAMETERS:                                                                *
!* (XC,YC): center of the ellipse; ALPHA: angle between two axes x' and x;    *
!* (A1,B1): semiaxis along the two ellipse axes (local) x' and y'             *
!* -------------------------------------------------------------------------- *

SUBROUTINE CHECK_AREA(areanum)

  IMPLICIT NONE
  REAL(8), INTENT(IN) :: areanum

  REAL(8), PARAMETER :: MYPI = 3.141592653589793238462643D0
  REAL(8), PARAMETER :: ALPHA = 0.48D0
  REAL(8), PARAMETER :: A1 = 0.170D0, B1 = 0.210D0
  REAL(8), PARAMETER :: XC = 0.523D0, YC = 0.475D0

  REAL(8) :: areana

  INTRINSIC DABS


  areana = MYPI*A1*B1

   write(*,*) '-----------------------------------------------------'
   write(*,*) '----------------- F: ellipse check ------------------'
   write(*,*) ' '
   write(*,100) areana
   write(*,101) areanum
   write(*,102) DABS(areanum-areana)
   write(*,103) DABS(areanum-areana)/areana
   write(*,*) '--------------- F: end ellipse check ----------------'
   write(*,*) '-----------------------------------------------------'
   write(*,*) ' '
   100 FORMAT('analytical area :', ES23.16)
   101 FORMAT('numerical area  :', ES23.16)
   102 FORMAT('absolute error  :', ES23.16)
   103 FORMAT('relative error  :', ES23.16)

  RETURN

END SUBROUTINE CHECK_AREA

!* -------------------------------------------------------------------------- *

REAL(8) FUNCTION IMPL_FUNC(xyz)

  IMPLICIT NONE

  REAL(8), DIMENSION(*), INTENT(IN) :: xyz

  REAL(8), PARAMETER :: MYPI = 3.141592653589793238462643D0
  REAL(8), PARAMETER :: ALPHA = 0.48D0
  REAL(8), PARAMETER :: A1 = 0.17D0, B1=0.21D0
  REAL(8), PARAMETER :: XC = 0.523D0, YC = 0.475D0

  REAL(8) :: x,y,a2,b2,ca,sa,c1,c2,c3,c4,c5,c6

  INTRINSIC DCOS,DSIN

  x = xyz(1)
  y = xyz(2)
  
  a2 = A1*A1
  b2 = B1*B1
  ca = DCOS(ALPHA)
  sa = DSIN(ALPHA)
  c1 = ca*ca/a2 + sa*sa/b2
  c2 = sa*sa/a2 + ca*ca/b2
  c3 = 2.D0*ca*sa*(b2-a2)/(a2*b2)
  c4 = -(2.D0*c1*XC + c3*YC)
  c5 = -(2.D0*c2*YC + c3*XC)
  c6 = 1.0D0 - (c1*XC*XC + c2*YC*YC + c3*XC*YC)
  
  IMPL_FUNC = c1*x*x + c2*y*y + c3*x*y + c4*x + c5*y - c6

END FUNCTION IMPL_FUNC

File: demo/C/2D/Ellipse/ellipse.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "ellipse.h"
typedef const double creal;
typedef const int cint;
typedef double real;

/* -------------------------------------------------------------------------- *
 * DESCRIPTION (reference phase where f(x,y) < 0):                            *
 * ellipse inside the square [0,1]x[0,1]                                      *
 * f(x,y) = c1*x^2 + c2*y^2 + c3*x*y + c4*x + c5*y - c6                       *
 * PARAMETERS:                                                                *
 * (XC,YC): center of the ellipse; ALPHA: angle between two axes x' and x;    *
 * (A1,B1): semiaxis along the two ellipse axes (local) x' and y'             *
 * -------------------------------------------------------------------------- */

int cont_line(real *xx,real *yy,cint ntot)
{
  int i;
  double dphi,phi,r0,ca,sa,xi,yi,a2,b2;

  ca = cos(ALPHA);
  sa = sin(ALPHA);
  a2 = A1*A1;
  b2 = B1*B1;

  dphi = 2.0*MYPI/(double)(ntot-1);

  for (i=1; i<ntot; i++) {
    phi = (i-1)*dphi;
    r0 = b2*a2/(b2*cos(phi)*cos(phi) + a2*sin(phi)*sin(phi));
    r0 = sqrt(r0);
    xi = r0*cos(phi);
    yi = r0*sin(phi);
    xx[i] = XC + xi*ca - yi*sa;
    yy[i] = YC + xi*sa + yi*ca;
  }

  xx[ntot] = xx[1];
  yy[ntot] = yy[1];

  return ntot;
}

/* -------------------------------------------------------------------------- */
void check_area(creal area_n)
{
  real area_a;
  
  area_a = MYPI*A1*B1;

  fprintf (stderr,"-----------------------------------------------------\n");
  fprintf (stderr,"------------------- Ellipse check -------------------\n\n");

  fprintf (stderr,"analytical area : %23.16e\n",area_a);
  fprintf (stderr,"numerical  area : %23.16e\n\n",area_n);
  fprintf (stderr,"absolute error  : %23.16e\n",fabs(area_a-area_n));
  fprintf (stderr,"relative error  : %23.16e\n",fabs(area_a-area_n)/area_a); 
  fprintf (stderr,"----------------- End ellipse check -----------------\n");
  fprintf (stderr,"-----------------------------------------------------\n\n");

  return;
}

/* -------------------------------------------------------------------------- */
real impl_func(creal xy[]) 
{
  real x,y,a2,b2,ca,sa,c1,c2,c3,c4,c5,c6,f0;

  x = xy[0];
  y = xy[1];

  a2 = A1*A1;
  b2 = B1*B1;
  ca = cos(ALPHA);
  sa = sin(ALPHA);
  c1 = ca*ca/a2 + sa*sa/b2;
  c2 = sa*sa/a2 + ca*ca/b2;
  c3 = 2.*ca*sa*(b2-a2)/(a2*b2);
  c4 = -(2.*c1*XC + c3*YC);
  c5 = -(2.*c2*YC + c3*XC);
  c6 = 1.0 - (c1*XC*XC + c2*YC*YC + c3*XC*YC);
  
  f0 = c1*x*x + c2*y*y + c3*x*y + c4*x + c5*y - c6;

  return f0;
}
